File: src/main/java/com/bdev/hogwarts_api/service/student/StudentServiceImpl.java
package com.bdev.hogwarts_api.service.student;

import com.bdev.hogwarts_api.dao.StudentDao;
import com.bdev.hogwarts_api.dao.StudentStatusDao;
import com.bdev.hogwarts_api.data.converter.student.StudentDtoConverter;
import com.bdev.hogwarts_api.data.converter.student.StudentModelConverter;
import com.bdev.hogwarts_api.data.dto.student.Student;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.Optional;

import static java.lang.String.format;
import static java.util.stream.Collectors.toList;

@Service
public class StudentServiceImpl implements StudentService {
    @Autowired
    private StudentDao studentDao;
    @Autowired
    private StudentStatusDao studentStatusDao;

    @Override
    public boolean exists(long studentId) {
        return studentDao.exists(studentId);
    }

    @Override
    public List<Student> getAllStudents() {
        return studentDao
                .findAll()
                .stream()
                .map(it -> StudentModelConverter.convert(
                        it,
                        studentStatusDao.findTopByStudentIdOrderByCreationTimeDesc(it.getId())
                ))
                .collect(toList());
    }

    @Override
    public Optional<Student> getStudentById(long studentId) {
        return Optional
                .ofNullable(studentDao.getOne(studentId))
                .map(it -> StudentModelConverter.convert(
                        it,
                        studentStatusDao.findTopByStudentIdOrderByCreationTimeDesc(it.getId())
                ));
    }

    @Override
    public long createStudent(Student student) {
        if (student.getId() != null) {
            throw new RuntimeException("Student id should be null during creation");
        }

        return studentDao.save(StudentDtoConverter.convert(student)).getId();
    }

    @Override
    public void editStudent(Student student) {
        if (!studentDao.exists(student.getId())) {
            throw new RuntimeException(format("Student with id '%d' does not exist", student.getId()));
        }

        studentDao.save(StudentDtoConverter.convert(student));
    }

    @Override
    public void deleteStudent(long studentId) {
        if (!studentDao.exists(studentId)) {
            throw new RuntimeException(format("Student with id '%d' does not exist", studentId));
        }

        studentDao.delete(studentId);
    }
}

File: src/main/java/com/bdev/hogwarts_api/service/student/StudentServiceImpl.kt
package com.bdev.hogwarts_api.service.student

import com.bdev.hogwarts_api.dao.StudentDao
import com.bdev.hogwarts_api.dao.StudentStatusDao
import com.bdev.hogwarts_api.data.converter.student.StudentDtoConverter
import com.bdev.hogwarts_api.data.converter.student.StudentModelConverter
import com.bdev.hogwarts_api.data.dto.student.Student
import com.bdev.hogwarts_api.data.model.student.StudentModel
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Service
import java.lang.String.format

@Service
class StudentServiceImpl : StudentService {
    @Autowired
    private lateinit var studentDao: StudentDao
    @Autowired
    private lateinit var studentStatusDao: StudentStatusDao

    override fun getAllStudents(): List<Student> {
        return studentDao
                .findAll()
                .map { getStudent(it) }
    }

    override fun getStudentById(studentId: Long): Student? {
        return studentDao
                .findOne(studentId)
                ?.let { getStudent(it) }
    }

    private fun getStudent(studentModel: StudentModel): Student {
        val id = studentModel.id ?: throw RuntimeException()

        return StudentModelConverter.convert(
                studentModel,
                studentStatusDao.findTopByStudentIdOrderByCreationTimeDesc(id) ?: throw RuntimeException()
        )
    }

    override fun exists(studentId: Long): Boolean {
        return studentDao.exists(studentId)
    }

    override fun createStudent(student: Student): Long {
        if (student.id != null) {
            throw RuntimeException("Student id should be null during creation")
        }

        return studentDao.save(StudentDtoConverter.convert(student)).id ?: throw RuntimeException()
    }

    override fun editStudent(student: Student) {
        if (!studentDao.exists(student.id)) {
            throw RuntimeException(format("Student with id '%d' does not exist", student.id))
        }

        studentDao.save(StudentDtoConverter.convert(student))
    }

    override fun deleteStudent(studentId: Long) {
        if (!studentDao.exists(studentId)) {
            throw RuntimeException(format("Student with id '%d' does not exist", studentId))
        }

        studentDao.delete(studentId)
    }
}
